File: win/system/backend/app/services/mental_state_service.py
"""
Mental state service
Manages system mental state tracking
"""

import json
from datetime import datetime, timezone
from typing import Optional
from app.core.config import MENTAL_STATE_FILE
from app.models import MentalState

def get_mental_state() -> MentalState:
    """Get current mental state from storage"""
    try:
        if MENTAL_STATE_FILE.exists():
            with open(MENTAL_STATE_FILE, "r") as f:
                state_data = json.load(f)
                # Convert the string back to datetime
                state_data["updated_at"] = datetime.fromisoformat(state_data["updated_at"])
                return MentalState(**state_data)
        else:
            # Return default state
            return MentalState(
                level="safe",
                updated_at=datetime.now(timezone.utc),
                notes=None
            )
    except Exception as e:
        print(f"Error loading mental state: {e}")
        return MentalState(
            level="safe",
            updated_at=datetime.now(timezone.utc),
            notes=None
        )

def save_mental_state(state: MentalState) -> bool:
    """Save mental state to storage"""
    try:
        state_data = state.dict()
        state_data["updated_at"] = state_data["updated_at"].isoformat()
        
        with open(MENTAL_STATE_FILE, "w") as f:
            json.dump(state_data, f, indent=2)
        
        return True
    except Exception as e:
        print(f"Error saving mental state: {e}")
        return False

def update_mental_state(level: str, notes: Optional[str] = None) -> MentalState:
    """
    Update mental state with new level and notes
    
    Args:
        level: Mental state level (safe, unstable, idealizing, self-harming, highly at risk)
        notes: Optional notes about the state
    
    Returns:
        Updated mental state
    """
    state = MentalState(
        level=level,
        updated_at=datetime.now(timezone.utc),
        notes=notes
    )
    
    save_mental_state(state)
    return state

File: co/system/backend/src/services/mental_state_service.ts
/**
 * Copyright (c) 2026 Clove Twilight
 * Licensed under the ESAL-2.0 Licence.
 * See LICENCE.md in the project root for full licence information.
 */

/**
 * Mental state service.
 * Manages system mental state tracking.
 */

import { existsSync } from 'node:fs';
import { readFile, writeFile } from 'node:fs/promises';

import { MENTAL_STATE_FILE } from '../core/config.js';
import type { MentalState } from '../models/index.js';
import { parseTimestamp } from '../utils/datetime.js';

function defaultState(): MentalState {
  return {
    level: 'safe',
    updated_at: new Date(),
    notes: null,
  };
}

/** Get current mental state from storage */
export async function getMentalState(): Promise<MentalState> {
  try {
    if (existsSync(MENTAL_STATE_FILE)) {
      const raw = await readFile(MENTAL_STATE_FILE, 'utf-8');
      const stateData = JSON.parse(raw) as { level: string; updated_at: string; notes?: string | null };

      return {
        level: stateData.level,
        updated_at: parseTimestamp(stateData.updated_at),
        notes: stateData.notes ?? null,
      };
    }

    // Return default state
    return defaultState();
  } catch (err) {
    console.error(`Error loading mental state: ${String(err)}`);
    return defaultState();
  }
}

/** Save mental state to storage */
export async function saveMentalState(state: MentalState): Promise<boolean> {
  try {
    const stateData = {
      level: state.level,
      updated_at: state.updated_at.toISOString(),
      notes: state.notes ?? null,
    };

    await writeFile(MENTAL_STATE_FILE, JSON.stringify(stateData, null, 2), 'utf-8');
    return true;
  } catch (err) {
    console.error(`Error saving mental state: ${String(err)}`);
    return false;
  }
}

/**
 * Update mental state with a new level and notes.
 *
 * @param level - Mental state level (safe, unstable, idealizing, self-harming, highly at risk)
 * @param notes - Optional notes about the state
 */
export async function updateMentalState(level: string, notes?: string | null): Promise<MentalState> {
  const state: MentalState = {
    level,
    updated_at: new Date(),
    notes: notes ?? null,
  };

  await saveMentalState(state);
  return state;
}